director/origin_monitor.go

```diff
@@ -0,0 +1,320 @@
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "Failed to start request for director test deletion")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode > 299 {
+		return errors.Errorf("Error response %v from director test deletion: %v", resp.StatusCode, resp.Status)
+	}
+
+	return nil
+}
+
+// Report the health status of test file transfer to origin
+func reportStatusToOrigin(originWebUrl string, status string, message string) error {
+	tkn, err := CreateDirectorTestReportToken(originWebUrl)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create a token for the diretor test upload")
+	}
+
+	reportUrl, err := url.Parse(originWebUrl)
+	if err != nil {
+		return errors.Wrap(err, "The origin URL is not parseable as a URL")
+	}
+
+	if status != "ok" && status != "error" {
+		return errors.Errorf("Bad status for reporting director test")
+	}
+
+	reportUrl.Path = "/api/v1.0/origin-api/directorTest"
+
+	dt := DirectorTest{
+		Status:    status,
+		Message:   message,
+		Timestamp: strconv.FormatInt(time.Now().Unix(), 10),
+	}
+
+	jsonData, err := json.Marshal(dt)
+	if err != nil {
+		// handle error
+		return errors.Wrap(err, "Failed to parse request body for reporting director test")
+	}
+
+	reqBody := bytes.NewBuffer(jsonData)
+
+	req, err := http.NewRequest("POST", reportUrl.String(), reqBody)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create POST request for reporting director test")
+	}
+
+	req.Header.Set("Authorization", "Bearer "+tkn)
+	req.Header.Set("Content-Type", "application/json")
+
+	tr := config.GetTransport()
+	client := http.Client{Transport: tr}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "Failed to start request for reporting director test")
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Wrap(err, "Failed to read response body for reporting director test")
+	}
+
+	if resp.StatusCode > 299 {
+		return errors.Errorf("Error response %v from reporting director test: %v", resp.StatusCode, string(body))
+	}
+
+	return nil
+}
+
+// Run a periodic test file transfer against an origin to ensure
+// it's talking to the director
+func PeriodicDirectorTest(originAd ServerAd) {
+	originName := originAd.Name
+	originUrl := originAd.URL.String()
+	originWebUrl := originAd.WebURL.String()
+
+	firstRound := true
+	for {
+		if firstRound {
+			time.Sleep(5 * time.Second)
+			log.Debug(fmt.Sprintf("Starting Director test for origin %s at %s", originName, originUrl))
+			firstRound = false
+		} else {
+			time.Sleep(15 * time.Second)
+		}
+		// End the periodic test if the ttlcache doesn't have the origin ad
+		// meaning the origin is offline and the cache is expired
+		serverAdMutex.RLock()
+		if !serverAds.Has(originAd) {
+			log.Debug(fmt.Sprintf("End of director test cycle for origin: %s at %s", originName, originUrl))
+			serverAdMutex.RUnlock()
+			return
+		}
+		serverAdMutex.RUnlock()
```

Comment:
Better to remove this and change to `defer serverAdMutex.RUnlock()`.  If anything in the control block above panics, you would end up with the lock held forever and potentially deadlock in the future.